lib/eknvfs/ekn-file.c

```diff
@@ -0,0 +1,326 @@
+ekn_file_equal (GFile *file1, GFile *file2)
+{
+  return g_str_equal (EKN_FILE_PRIVATE (file1)->uri, EKN_FILE_PRIVATE (file2)->uri);
+}
+
+static gboolean
+ekn_file_is_native (GFile *self)
+{
+  return TRUE;
+}
+
+static gboolean 
+ekn_file_has_uri_scheme (GFile *self, const char *uri_scheme)
+{
+  return g_str_equal ("ekn", uri_scheme);
+}
+
+static char *
+ekn_file_get_uri_scheme (GFile *self)
+{
+  return g_strdup ("ekn");
+}
+
+static char *
+ekn_file_get_basename (GFile *self)
+{
+  EknFilePrivate *priv = EKN_FILE_PRIVATE (self);
+  return priv->uri ? g_path_get_basename (priv->uri + EKN_SCHEME_LEN) : NULL;
+}
+
+static char *
+ekn_file_get_path (GFile *self)
+{
+  EknFilePrivate *priv = EKN_FILE_PRIVATE (self);
+  return priv->uri ? g_strdup (priv->uri + EKN_SCHEME_LEN) : NULL;
+}
+
+static char *
+ekn_file_get_uri (GFile *self)
+{
+  return g_strdup (EKN_FILE_PRIVATE (self)->uri);
+}
+
+static char *
+ekn_file_get_parse_name (GFile *self)
+{
+  return g_strdup (EKN_FILE_PRIVATE (self)->uri);
+}
+
+static GFile *
+ekn_file_get_parent (GFile *self)
+{
+  return NULL;
+}
+
+static GFileInfo *
+ekn_file_query_info (GFile                *self,
+                     const char           *attributes,
+                     GFileQueryInfoFlags   flags,
+                     GCancellable         *cancellable,
+                     GError              **error)
+{
+  EknFilePrivate *priv = EKN_FILE_PRIVATE (self);
+  GFileAttributeMatcher *matcher;
+  GFileInfo *info;
+
+  info    = g_file_info_new ();
+  matcher = g_file_attribute_matcher_new (attributes);
+
+  if (g_file_attribute_matcher_matches (matcher, G_FILE_ATTRIBUTE_STANDARD_SIZE))
+    g_file_info_set_size (info, eos_shard_blob_get_content_size (priv->blob));
+
+  if (g_file_attribute_matcher_matches (matcher, G_FILE_ATTRIBUTE_STANDARD_CONTENT_TYPE))
+    g_file_info_set_content_type (info, eos_shard_blob_get_content_type (priv->blob));
+
+  g_file_attribute_matcher_unref (matcher);
+
+  return info;
+}
+
+static GFileInputStream *
+ekn_file_read_fn (GFile *self, GCancellable *cancellable, GError **error)
+{
+  GInputStream *stream = eos_shard_blob_get_stream (EKN_FILE_PRIVATE (self)->blob);
+  return _ekn_file_input_stream_wrapper_new (self, stream);
+}
+
+static void
+ekn_file_iface_init (gpointer g_iface, gpointer iface_data)
+{
+  GFileIface *iface = g_iface;
+
+  iface->dup            = ekn_file_dup;
+  iface->hash           = ekn_file_hash;
+  iface->equal          = ekn_file_equal;
+  iface->is_native      = ekn_file_is_native;
+  iface->has_uri_scheme = ekn_file_has_uri_scheme;
+  iface->get_uri_scheme = ekn_file_get_uri_scheme;
+  iface->get_basename   = ekn_file_get_basename;
+  iface->get_path       = ekn_file_get_path;
+  iface->get_uri        = ekn_file_get_uri;
+  iface->get_parse_name = ekn_file_get_parse_name;
+  iface->get_parent     = ekn_file_get_parent;
+  iface->query_info     = ekn_file_query_info;
+  iface->read_fn        = ekn_file_read_fn;
+}
+
+static GType
```

Comment:
This cannot be a private GType, otherwise you won't be able to access it as an actual type, but only as a `GBoxed`.